domain/keymanager/state/state.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/user"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/keymanager"
+	keyerrors "github.com/juju/juju/domain/keymanager/errors"
+	jujudb "github.com/juju/juju/internal/database"
+)
+
+// State represents a type for interacting with the underlying state of a user's
+// public key in a model.
+type State struct {
+	*domain.StateBase
+}
+
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// AddPublicKeyForUser is responsible for adding one or more ssh public keys for
+// a user. If no keys are provided to the operation then nothing will happen and
+// no error produced. The following errors can be expected.
+// [keyerrors.PublicKeyAlreadExists] - When one of the public keys being added
+// for a user already exists.
+func (s *State) AddPublicKeysForUser(
+	ctx context.Context,
+	userId user.UUID,
+	publicKeys []keymanager.PublicKey,
+) error {
+	if len(publicKeys) == 0 {
+		return nil
+	}
+
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertStatement, err := s.Prepare(`
+INSERT INTO user_public_ssh_key (comment,
+                                 fingerprint,
+						         public_key,
+                                 user_id,
+                                 fingerprint_hash_algorithm_id)
+SELECT $userPublicKeyInsert.comment,
+       $userPublicKeyInsert.fingerprint,
+       $userPublicKeyInsert.public_key,
+       $userPublicKeyInsert.user_id,
+       s.id	
+FROM ssh_fingerprint_hash_algorithm s
+WHERE s.algorithm = $userPublicKeyInsert.algorithm
+`, userPublicKeyInsert{})
+
+	if err != nil {
+		return fmt.Errorf(
+			"preparing insert statement for adding user %q public keys: %w",
+			userId,
+			domain.CoerceError(err),
+		)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// We can't perform a bulk insert here because of the foreign key lookup
+		// for the algorithm. It could be done with a temp table but this is not
+		// considered a hot path.
```

Comment:
You could select the algorithm ids upfront and then do a bulk insert, but like you say, it's not a hot path and probably not worth it.

Reply:
Yeah small fry at the moment I think.